### ISSUE TITLE
BUG: pipx upgrade doesn't understand extra dependencies

### ISSUE THREAD

**Issue by neutrinoceros** [pypa/pipx#925]
**Describe the bug**

This works correctly
```
pipx install "mycli[extras]"
```

but this doesn't
```
pipx install "mycli[extras]"
pipx upgrade "mycli[extras]"
```
and it fails with
```
Package is not installed. Expected to find /path/to/pipx/venvs/mycli[isolated], but it does not exist.
```

**How to reproduce**
For reproduction, I suggest

```
pipx install "coverage[toml]"
pipx upgrade "coverage[toml]"
```

**Expected behavior**
I'm not 100% sure what should happen actually, but the current error message seems broken (*of course* a directory with `[` and `]` in its name wasn't found).
I think the upgrade should go through wether or not the extra targets match the ones I used (or didn't use) the first time I installed the CLI (here `coverage`).

I'd be happy to contribute a patch, but I'd appreciate some guidance regarding what the expected behaviour should be.

**Comment by meowmeowmeowcat** [comment]
You can simply run `pipx upgrade mycli` and you don't need to pass extra dependencies since it's installed. But I think we can improve the error message to make it more clearly.

**Comment by uranusjr** [comment]
Does `pipx upgrade mycli` upgrade dependencies that are only pulled in by the extra? I’m under the impression it only tries to upgrade the specified package (i.e. `mycli` here); if so we would actually need to add some additional logic for the extra-ed case.

**Comment by meowmeowmeowcat** [comment]
> Does pipx upgrade mycli upgrade dependencies that are only pulled in by the extra?

Not really, unless `--pip-args='--upgrade-strategy eager'` is passed. This is probably related to #254.

**Comment by uranusjr** [comment]
Even if we change the strategy, dependencies in the extra would still not be upgraded (because from pip’s perspective those packages are unrelated unless you pass in the extra). So for now `pipx upgrade` does not have a difference in behaviour whether you pass in the extra or not, but would if we change the default upgrade strategy. I’d say the current behaviour is OK (barring a better error message), but we should probably change the strategy (or have a flag to control it).

**Comment by meowmeowmeowcat** [comment]
> Even if we change the strategy, dependencies in the extra would still not be upgraded (because from pip’s perspective those packages are unrelated unless you pass in the extra)

Looks like pipx will still run `pip install --upgrade mycli[extras]` for upgrading even if the command is `pipx upgrade mycli`, so it also upgrades extras at the same time actually.

**Comment by layday** [comment]
> [...] which would requite us to also fix how extras are passed to `upgrade`.

This might've been fixed since? It looks like if I install `pipx install nox[uv]` then `pipx upgrade nox` correctly issues `pip --no-input install --upgrade nox[uv]`.  Therefore, the only change required is to pass `--upgrade-strategy=eager`.

**Comment by cjolowicz** [comment]
What should pipx do if the main app's version didn't change? Should it still upgrade dependencies?

(IMO it should. This will let users benefit from bugfixes and security updates to the app's dependency tree. It will also help with the more exotic case of `nox[uv]` where uv has a much faster release cycle than Nox.)

**Comment by lazka** [comment]
I've just checked that "uv" does it correctly. If I install a tool, and I run "uv tool upgrade <tool>" it updates the transitive dependencies as expected.

**Comment by Vrihub** [comment]
What about this use case: package "x" is installed, but its latest release also requires the "y" extra for proper functioning from now on (yt-dlp anyone?); user only installed "x" and now wants to upgrade "x" AND install the "y" extra too.

Running `pipx upgrade 'x[y]'` will fail with the usual message ("Expected to find /path/to/pipx/venvs/x[y]"), even with `--force`.

To get the desired behaviour, one must run `pipx install --force 'x[y]'`, maybe this workaround should be suggested somewhere in the docs (or even in the error message).

**Comment by ncoghlan** [comment]
From a docs perspective, it would be good to at least cover the `pipx upgrade --pip-args="--upgrade-strategy=eager" app-package` workaround (and the equivalent for `pipx upgrade-all`).

Trying that workaround (which does indeed work!), I also noticed that if the original installation relied on `--pip-args` for correctness (e.g. passing `--prefer-binary`), those arguments are NOT preserved and need to be specified again when running the update.

On that front, should this issue be retitled? `pipx upgrade` remembers the originally specified extras now, but:

* it doesn't update the transitive dependencies by default
* it forgets the original `--pip-args` (and hence may fail if those args were necessary to make the installation work)
